File: scripts/validateImageRequest.py
# Verify that image requests are made by logged in users.
#
# This AWS lambda handler is called for a CloudFront User Access event
# (not the Origin access event) which happens early in the request
# process.

import json
import base64
from datetime import datetime
import json
import urllib.request
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.serialization import load_pem_public_key
# Lambda@edge doesn't support importing external libraries without
# extra work. You need to deploy your code and its dependencies
# together in a zip file. The maximum size for a .zip deployment
# package for Lambda is 250 MB (unzipped).  Lambda@Edge doesn't
# support layers or containers.
# See:
# * https://docs.aws.amazon.com/lambda/latest/dg/python-package.html
# * https://stackoverflow.com/questions/77986827/cloudfront-lambdaedge-in-nodejs-layer-is-not-supported-for-cloudfront-how-to

# Global variables can be used for caching because AWS Lambda reuses
# the same container for multiple invocations. When the container is
# first started and the handler is called, it's known as a "cold
# start". Subsequent calls to the already running handler are known as
# "warm starts". You can also cache static assets locally in the /tmp
# directory. The AWS Lambda SDK for JavaScript reuses TCP connections
# by default. What about python? The keepAliveMsecs default is 1000
# ms.  todo: what is the keepAliveMseces for Python?

# Cache the keys (jwks) so we only have to fetch them on a cold start
# or when the keys change.
cachedJwks = None

# Get the userPoolId and client_id from aws_settings.json.
userPoolId = "us-west-2_4czmlJC5x"
client_id = "47ahgb3e4jqhk86o7gugvbglf8"

def lambda_handler(event, context):
  """
  Verify that image requests are made by logged in users.

  The event parameter is json that contains variables about the
  request. You can determine the request url and the header
  information from it.

  The context parameter is an object that tells about the running
  environment.

  Return a requests dictionary.
  """
  # You can log by printing. You can see logging when using the lambda
  # testing tab, however, requests from the internet are not logged
  # unless you enable it in cloudfront.

  # print("Lambda handler event: ", json.dumps(event))

  # Get the request url and the access_token in the header.
  request = event['Records'][0]['cf']['request']
  url = request['uri']
  headers = request['headers']
  if 'auth' in headers:
    access_token = headers['auth'][0]['value']
  else:
    access_token = None

  # For image requests deny access if the user is not logged in. Do
  # this by checking the url and by validating the access token.
  try:
    message = validateImageRequest(url, access_token)
    if message:
      print(message)
  except Exception as ex:
    print(str(ex))
    message = "Got an unexpected exception."
  if message:
    return {
      'status': '403',
      'statusDescription': 'Forbidden',
      'body': 'Unauthorized access.',
    }
  return request

def validateImageRequest(url, access_token):
  """
  Verify that image requests are made by logged in users.

  Return an empty string when valid, else return a message telling
  what went wrong. It may raise an exception for unexpected cases.
  """
  # Allow all non-image requests.
  if "images/" not in url:
    return ""

  global cachedJwks
  if cachedJwks:
    jwks = cachedJwks
  else:
    print("No cached jwks, cold start.")
    jwks = fetchJwks()
    if jwks is None:
      return "Unable to fetch the jwks."
    cachedJwks = jwks

  return validateAccessToken(jwks, access_token)

def fetchJwks():
  """
  Fetch the Cognito signing keys (jwks) using https and return
  them as a dictionary.
  """
  region = userPoolId.split("_")[0]
  jwksUrl = f"https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/jwks.json"
  with urllib.request.urlopen(jwksUrl) as response:
    if response.status != 200:
      raise Exception("Unable to fetch the jwks.")
    return json.loads(response.read().decode())

def findJwksKey(kid, jwks):
  """
  Find a key in the jwks list given the key id. Return the key or
  None when not found.
  """
  for k in jwks["keys"]:
    if k['kid'] == kid:
      return k
  return None

def base64_url_decode(data):
  padding = '=' * (4 - len(data) % 4)
  return base64.urlsafe_b64decode(data + padding)

def validateAccessToken(jwks, token):
  """
  Validate the access token.

  Return an empty string when valid, else return a message telling
  what went wrong. It may raise an exception for unexpected cases.
  """
  if token is None:
    return "No token passed."

  header_b64, payload_b64, signature_b64 = token.split('.')
  header = json.loads(base64_url_decode(header_b64).decode('utf-8'))
  payload = json.loads(base64_url_decode(payload_b64).decode('utf-8'))
  signature = base64_url_decode(signature_b64)

  # The jwk keys get rotated about every 6 months. If the key is not
  # found, fetch new ones and look again.
  key = findJwksKey(header["kid"], jwks)
  if key is None:
    jwks = fetchJwks()
    key = findJwksKey(header["kid"], jwks)
    if key is None:
      return "Signing key not found."
    print("Keys rotated.")
    global cachedJwks
    cachedJwks = jwks

  # Cognito only uses RS256 and that is the only algorithm this code
  # supports.
  if header["alg"] != "RS256":
    return "Unsupported header algorithm."
  if key["alg"] != "RS256":
    return "Unsupported key algorithm."

  # Construct the RSA public key.
  n = int.from_bytes(base64_url_decode(key['n']), 'big')
  e = int.from_bytes(base64_url_decode(key['e']), 'big')
  public_key = rsa.RSAPublicNumbers(e, n).public_key()

  # Verify the signature.
  try:
    public_key.verify(
      signature,
      f"{header_b64}.{payload_b64}".encode('utf-8'),
      padding.PKCS1v15(),
      hashes.SHA256()
    )
  except Exception as e:
    return "Invalid signature."

  if payload.get('client_id') != client_id:
    return "Invalid client id."

  region = userPoolId.split("_")[0]
  iss = f"https://cognito-idp.{region}.amazonaws.com/{userPoolId}"
  if payload.get('iss') != iss:
    return "Invalid iss."

  # Validate expiration
  if payload.get('exp') < datetime.utcnow().timestamp():
    return "Token expired."

  return ""

File: scripts/validateImageRequest.js
"use strict";

const https = require('https');
const jwt = require('jsonwebtoken');
const jwkToPem = require('jwk-to-pem');

// Set these values from the values in aws-settings.json.
const userPoolId = "us-west-2_4czmlJC5x"
const client_id = '47ahgb3e4jqhk86o7gugvbglf8'

const region = userPoolId.split("_")[0]
const jwksUrl = `https://cognito-idp.${region}.amazonaws.com/${userPoolId}/.well-known/jwks.json`;
const iss = `https://cognito-idp.${region}.amazonaws.com/${userPoolId}`

// Cache the keys (jwks) so we only have to fetch them on a cold start.
let jwks = null

async function getJwks() {
  // Fetch the cognito public keys used to encrypt tokens.
  return new Promise((resolve, reject) => {
    https.get(jwksUrl, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => resolve(JSON.parse(data).keys));
      res.on('error', reject);
    });
  });
}

function parseQueryString(qs) {
  // Function to parse the request query string into an object.
  if (qs == undefined)
    return {}
  const params = qs.split('&');
  const result = {};
  for (const param of params) {
    if (param === "") continue
    let [key, value] = param.split('=');
    if (key === "" || value == "") continue
    result[decodeURIComponent(key)] = value;
  }
  return result;
}

async function verifyJwt(token, ignoreExpiration) {
  // Verify the token is valid for our cognito user.

  if (!token) {
    throw new Error('Missing token.');
  }
  if (ignoreExpiration == undefined) {
    throw new Error('Missing ignoreExpiration.');
  }

  // Decode JWT header.
  const { header } = jwt.decode(token, { complete: true });
  if (!header || !header.kid || !header.alg)
    throw new Error('Invalid token.');

  // If we don't have the keys (jwks), get and cache them.
  if (jwks === null) {
    console.log("jwks: Cold start.")
    jwks = await getJwks();
  }
  else {
    console.log("jwks: Warm start.")
  }

  // Find the key used to encrypt the token.
  const key = jwks.find((key) => key.kid === header.kid);
  if (key) {
    console.log(`key: ${key.kid}`)
  }
  else {
    // When the keys roll over there is a transition period where the
    // old keys still exist. As long as this code gets a cold start in that
    // period, the new keys get cached.
    throw new Error('Signing key not found.');
  }

  if (key.kty != "RSA") {
    throw new Error(`Algorithm not supported. key.kty: ${key.kty}`);
  }

  // Convert the jwk key to a PEM public key.
  var pem = jwkToPem(key);

  // Verify that the token was for our cognito pool and the RS256
  // algorithm was used.
  const options = {
    algorithms: ['RS256'],
    issuer: iss,
  }

  // Ignore expiration when testing.
  if (ignoreExpiration === true) {
    console.log("warning: Ignoring token expiration for testing.")
    options['ignoreExpiration'] = ignoreExpiration
  }

  // Verify the token was created with the Cognito public key.
  // See https://www.npmjs.com/package/jsonwebtoken for docs.
  const payload = jwt.verify(token, pem, options)

  if (payload.client_id != client_id)
    throw new Error('Wrong client_id.')

  // Log the expire date when we are ignoring it.
  if (ignoreExpiration === true) {
    const expires = new Date(payload.exp * 1000)
    const dateString = expires.toISOString()
    const formattedString = dateString.replace(/[TZ]/g, ' ');
    console.log(`Expires: ${formattedString}`);
  }

  return payload;
}

async function validateRequest(url, token, id, user, ignoreExpiration) {
  // Return true when the request is valid.

  // Allow all non-image requests.
  if (!url.startsWith('/images/')) {
    return true
  }

  // Make sure the id and user exist.
  if (!id && !user) {
    console.error("No query parameters.");
    return false
  }
  if (!user) {
    console.error("Missing user query parameter.")
    return false
  }
  if (!id) {
    console.error("The id query parameter is missing.")
    return false
  }

  // Make sure the token exists.
  if (!token) {
    // Most likely this is a direct request outside the website for an
    // image.  If you get this, it could be because you need to allow
    // auth header in Cloudfront Cache policy.
    console.error("No token in the auth header.")
    return false
  }

  // Validate the token.
  let payload = {}
  try {
    payload = await verifyJwt(token, ignoreExpiration);
  }
  catch (err) {
    console.error(`VerifyJwt: ${err.message}`);
    return false
  }

  // Validate that the user name on the url is the same as the token
  // user name.
  if (user != payload.username) {
    console.error("Users do not match.")
    console.log(`Url user: ${user}`)
    console.log(`Token user: ${payload.username}`)
    return false
  }

  return true
}

async function handler(event, context) {
  // Validate the request. For image requests make sure the user is
  // logged in.

  // Make sure this code is handling the viewer request event not some
  // other event.
  const cf = event.Records[0].cf
  if (cf.config && cf.config.eventType) {
    const eventType = cf.config.eventType
    if (eventType && eventType != "viewer-request") {
      console.error(`This code is connected to the wrong event: ${eventType}`)
      return {
        'status': '400',
        'statusDescription': 'Bad Request',
      }
    }
  }

  const request = cf.request;
  const url = request.uri
  const headers = request['headers']
  const queryParams = parseQueryString(request.querystring);
  const id = queryParams.id

  console.log(`url: ${url}`);
  console.log(`id: ${queryParams.id}`);
  console.log(`user: ${queryParams.user}`);

  // Get the access token from the auth header. It only exists on
  // image requests.
  let access_token = ""
  if (headers.auth)
    access_token = headers.auth[0].value

  // todo: change this to false after testing.
  const ignoreExpiration = true

  const ok = await validateRequest(url, access_token, queryParams.id,
                                   queryParams.user, ignoreExpiration)

  if (ok) {
    console.log("auth: Passed");
    return request;
  }
  else {
    console.error("auth: Failed");
    return {
      'status': '401',
      'statusDescription': 'Unauthorized',
    }
  }
}

module.exports = { getJwks, verifyJwt, handler, region, userPoolId, client_id, iss,
                   parseQueryString, validateRequest};
